### PR TITLE
Add the ability to navigate to the previously displayed page

### PR DESCRIPTION
**Pull request by Joppe27**
Hi, I found your extension a few months ago and it's been very useful for me. There was just one aspect missing to really make it perfect for my use-case, which is the ability to navigate back in the history of pages you have visited. I have added that functionality in this PR as an option for the extension icon's behavior.

Note:
- This functionality does not work using a combination of Google and the 'Default' user agent, because Google appends tracking information to links which breaks navigation. I personally don't use the 'Default' user agent as it doesn't properly resize content so this is not a problem for me, but definitely let me know what you think about this limitation. This PR works fine using Google via all 3 other user agents.
- I don't speak Japanese and have as such used an online tool for `navigateBack`'s translation. Please check whether it is correct, as it likely won't be.

Lastly, you're naturally free to close this PR if you don't accept outside contributions or don't have the time to review them. Thanks!

### CODE REVIEW

**Comment by Soushi-Atsumi** [comment]
Sorry to be late for replying. I will check the pull request.

**Comment by Soushi-Atsumi** [comment]
> This functionality does not work using a combination of Google and the 'Default' user agent, because Google appends tracking information to links which breaks navigation.

In my environment, it works fine even if I use the Default user agent. The behavior may depend on the environment of a user.

**Comment by Soushi-Atsumi** [comment]
In my environment, it works fine even if I use the Default user agent. The behavior may depend on the environment of a user.

File: options/options.html [hunk 1]
```diff
@@ -68,6 +68,7 @@
 		<fieldset>
 			<legend id="pageActionLegend"></legend>
 			<input type="radio" id="pageActionReloadInputRadio" name="pageAction" checked="checked"><label id="pageActionReloadLabel"></label><br>
+			<input type="radio" id="pageActionNavigateBackInputRadio" name="pageAction" checked="checked"><label id="pageActionNavigateBackLabel"></label><br>
```
Comment (Soushi-Atsumi) [inline options/options.html]: Remove "checked="checked"".

File: background/background.js [hunk 2]
```diff
@@ -79,7 +81,19 @@ async function main() {
...
+
+		// Clear navigation history when re-opening the sidebar for logical UX and to avoid issues.
+		previousPages = [`${searchEngine}${searchEngineQuery}${selectionText}`];
 	});
+
+	browser.webRequest.onBeforeSendHeaders.addListener(savePreviousPage, backNavigatableUrls);
+	function savePreviousPage(details) {
+		// Disallow navigating back to internal extension URLs as it causes issues and is pointless.
+		if (details.originUrl.includes("moz-extension") === false) {
+			previousPages.push(details.originUrl);
```
Comment (Soushi-Atsumi) [inline background/background.js]: This conflicts with other extension (e.g. [Open in Sidebar](https://github.com/Soushi-Atsumi/Open-in-Sidebar#open-in-sidebar)) which loads a webpage. When other extension loads a webpage, this extension saves the webpage to the history.

**Comment by Joppe27** [comment]
I've just checked this again on my computer and it now works fine here for all user agents too. This means something has probably changed in the meantime at Google's end, which is good news.

I will do the following:
- Remove the option's disclaimer in parentheses since it's not necessary anymore.
- I would also like to change my wording to better reflect the option's behavior. I propose "Go back to search results" instead of the current "Go back to the previous page" as that's not really what the option does.

```diff --git a/_locales/en/messages.json b/_locales/en/messages.json
--- a/_locales/en/messages.json	(revision 7ec5e0cc19f021cb1de1d4b6957bf1b2c643863d)
+++ b/_locales/en/messages.json	(date 1689533490721)
@@ -126,7 +126,7 @@
 	"navigateBack": {
-		"message": "Navigate back to previous page (does not work on 'Default' user agent)"
+		"message": "Navigate back to search results"
 	},
===================================================================
--- a/_locales/ja/messages.json	(revision 7ec5e0cc19f021cb1de1d4b6957bf1b2c643863d)
+++ b/_locales/ja/messages.json	(date 1689536192502)
@@ -126,7 +126,7 @@
 	"navigateBack": {
-		"message": "前のページに戻る（「既定値」 ユーザエージェントでは動作しません。）"
+		"message": "検索結果に戻る"
 	},
```
Just give me a heads up if the above looks good and I'll apply the changes (please check the Japanese translation, I used machine translation for this again so it might be incorrect). If there's anything else that's required for this to be merged let me know and I'll look into it.

**Comment by Soushi-Atsumi** [comment]
"Go back to search results" is incorrect in some cases. For example, number three goes back to number two which is not a page of search results in the following case.

1. https://www.google.com/search?q=mdn (search results)
2. https://developer.mozilla.org/docs/Web
3. https://developer.mozilla.org

So I think "Navigate back to previous page" is better.

**Comment by Joppe27** [comment]
Please wait a few days while I investigate the inconsistent behavior you mentioned. It might be a bug in my code.

**Comment by Joppe27** [comment]
I've applied some changes to hopefully make my code more reliable and also to improve UX (info in commit messages). Please let me know if you're still experiencing issues.

File: options/options.html [hunk 3]
```diff
@@ -68,7 +68,9 @@
 		<fieldset>
 			<legend id="pageActionLegend"></legend>
 			<input type="radio" id="pageActionReloadInputRadio" name="pageAction" checked="checked"><label id="pageActionReloadLabel"></label><br>
-			<input type="radio" id="pageActionGoBackToHomeInputRadio" name="pageAction"><label id="pageActionGoBackToHomeLabel"></label>
+			<input type="radio" id="pageActionGoBackToHomeInputRadio" name="pageAction"><label id="pageActionGoBackToHomeLabel"></label><br><br>
```
Comment (Soushi-Atsumi) [inline options/options.html]: Change "<br&gt;<br&gt;" to "<br&gt;". I think it is better.

File: options/options.js [hunk 4]
```diff
@@ -26,6 +26,7 @@ const searcheEngineGoogleId = 'searcheEngineGoogleInputRadio';
 const searcheEngineYahooId = 'searcheEngineYahooInputRadio';
 const searcheEngineYahooJapanId = 'searcheEngineYahooJapanInputRadio';
 const pageActionReloadId = 'pageActionReloadInputRadio';
+const pageActionNavigateBackId = 'pageActionNavigateBackInputRadio';
```
Comment (Soushi-Atsumi) [inline options/options.js]: Put this after "const pageActionGoBackToHomeId = 'pageActionGoBackToHomeInputRadio';".

File: options/options.js [hunk 5]
```diff
@@ -150,6 +151,9 @@ function checkPageAction() {
 			case pageActions.reload:
 				document.getElementById(pageActionReloadId).checked = true;
 				break;
+			case pageActions.navigateBack:
+				document.getElementById(pageActionNavigateBackId).checked = true;
+				break;
```
Comment (Soushi-Atsumi) [inline options/options.js]: Put this after "	case pageActions.goBackToHome:".

File: options/options.js [hunk 6]
```diff
@@ -203,13 +207,15 @@ function initDocuments() {
...
 	document.getElementById('pageActionReloadLabel').innerText = browser.i18n.getMessage('reload');
+	document.getElementById('pageActionNavigateBackLabel').innerText = browser.i18n.getMessage('navigateBack');
 	document.getElementById('pageActionGoBackToHomeLabel').innerText = browser.i18n.getMessage('goBackToHome');
 	document.getElementById('additionalPermissionsLegend').innerText = browser.i18n.getMessage('additionalPermissions');
 	document.getElementById('hostLabel').innerText = browser.i18n.getMessage('host');
 	document.getElementById('useragentLegend').innerText = browser.i18n.getMessage('useragent');
 	document.getElementById('defaultLabel').innerText = browser.i18n.getMessage('default');
-	document.getElementById('informationDivision').innerText = browser.i18n.getMessage('optionsUserAgentHTMLInformation');
-	document.getElementById('cautionDivision').innerText = browser.i18n.getMessage('optionsUserAgentHTMLCaution');
+	document.getElementById('navigateBackInformationDivision').innerText = browser.i18n.getMessage('optionsHostHTMLInformation');
```
Comment (Soushi-Atsumi) [inline options/options.js]: The message optionsHostHTMLInformation leads misunderstanding. A new message for this feature (e.g. "Navigate back requires the host permission.") should be added.

File: options/options.js [hunk 7]
```diff
@@ -231,6 +237,9 @@ function pageActionRadioButtonOnClick(event) {
 		case pageActionReloadId:
 			saveConfig({ [storageKeys.pageAction]: pageActions.reload });
 			break;
+		case pageActionNavigateBackId:
+			saveConfig({ [storageKeys.pageAction]: pageActions.navigateBack });
+			break;
```
Comment (Soushi-Atsumi) [inline options/options.js]: Put this after "case pageActionGoBackToHomeId:".

File: options/options.js
Comment (Soushi-Atsumi) [inline options/options.js:1]: The config of storageKeys.pageAction is still pageActions.navigateBack when the host permission has been removed. It should be changed to pageActions.reload when the permission has been removed.

**Review by Soushi-Atsumi — CHANGES_REQUESTED** [review]

File: options/options.html (hunk 1 above)
Comment (Joppe27) [reply options/options.html]: fbe1aae

File: background/background.js (hunk 2 above)
Comment (Joppe27) [reply background/background.js]: 47ecc5a

File: options/options.html (hunk 3 above)
Comment (Joppe27) [reply options/options.html]: fbe1aae

File: options/options.js (hunk 4 above)
Comment (Joppe27) [reply options/options.js]: fbe1aae

File: options/options.js (hunk 5 above)
Comment (Joppe27) [reply options/options.js]: fbe1aae

File: options/options.js (hunk 6 above)
Comment (Joppe27) [reply options/options.js]: 0143a47

File: options/options.js (hunk 7 above)
Comment (Joppe27) [reply options/options.js]: fbe1aae

File: options/options.js
Comment (Joppe27) [reply options/options.js:1]: 5be2f42

File: _locales/en/messages.json [hunk 8]
```diff
@@ -125,10 +125,18 @@
...
+		"message": "Navigate back to previous page"
+	},
+
 	"optionsHTMLTitle": {
 		"message": "Options",
 		"description": "The title of options.html."
 	},
+	
+	"optionsNavigateBackHTMLInformation": {
+		"message": "The \"navigate back to previous page\" feature requires the host permission."
```
Comment (Soushi-Atsumi) [inline _locales/en/messages.json]: "The \"Navigate back to previous page\" feature requires the host permission." is a bit better.

File: options/options.js [hunk 9]
```diff
@@ -373,8 +382,16 @@ function searchEngineWhichHasBeenAddedDeleteButtonOnClick(event) {
 	}
 }
 
-function toggleUserAgentRadioDisabled(disabled) {
+function toggleHostControlledRadioDisabled(disabled) {
 	document.options.userAgent.forEach(element => element.disabled = disabled);
+	
+	if (disabled && document.getElementById("pageActionNavigateBackInputRadio").checked)
+	{
+		document.getElementById("pageActionNavigateBackInputRadio").checked = false;
```
Comment (Soushi-Atsumi) [inline options/options.js]: This line is unnecessary.

File: background/background.js [hunk 10]
```diff
@@ -79,7 +84,47 @@ async function main() {
...
+		// local GUID of Search in Sidebar. If there's no match, the request is coming from a different extension which
+		// means it doesn't need to be saved. It also means Search in Sidebar is closed, which means we can keep
+		// ignoring requests until it is opened again (which is why this is a global var).
+		if (details.originUrl.includes("moz-extension") && details.originUrl !== browser.runtime.getURL('')) {
+			externalSidePanelOpen = true;
+			return;
+		}
+		
+			// Disallow saving Search in Sidebar's origin URL as it causes issues and is pointless.
+		if (details.originUrl.includes("moz-extension") === false &&
```
Comment (Soushi-Atsumi) [inline background/background.js]: [String.prototype.includes()](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/String/includes) always returns boolean so ' === false' is unnecessary.

File: background/background.js [hunk 11]
```diff
@@ -25,6 +29,7 @@ const mainAdditionalSearchEngineMenuItemId = 'mainAdditionalSearchEngine';
 const yahooMenuItemId = 'yahoo';
 const yahooJapanMenuItemId = 'yahooJapan';
 const hostPermissions = { origins: ['*://*/*'] };
+const backNavigatableUrls = { urls: ['*://*/*'], types: ['main_frame'] };
```
Comment (Soushi-Atsumi) [inline background/background.js]: Add 'tabId: -1' to the filter. Without this filter option, the event listener is called when a tab loads a page.

File: background/background.js (hunk 2 above)
Comment (Soushi-Atsumi) [reply background/background.js]: externalSidePanelOpen is true when a non sidebar extension (e.g. [Open-in-Popup](https://github.com/Soushi-Atsumi/Open-in-Popup)) loads a webpage even if the sidebar for this extension is opened.

**Review by Soushi-Atsumi — CHANGES_REQUESTED** [review]

File: background/background.js (hunk 10 above)
Comment (Joppe27) [reply background/background.js]: My IDE complains about this but you're absolutely right. Have applied in 6566d78.

File: background/background.js (hunk 11 above)
Comment (Joppe27) [reply background/background.js]: 6566d78

File: options/options.js (hunk 9 above)
Comment (Joppe27) [reply options/options.js]: b3b9903

File: _locales/en/messages.json (hunk 8 above)
Comment (Joppe27) [reply _locales/en/messages.json]: f488bed

File: background/background.js (hunk 10 above)
Comment (Soushi-Atsumi) [reply background/background.js]: So you don't need to put a boolean at the right side. Simply write like: `if (details.originUrl.includes("moz-extension") &&`.

Comment (Joppe27) [reply background/background.js]: Did you mean `!details.originUrl.includes("moz-extension")`? If not, sorry but I don't understand. If so, do you want me to change all cases of `something === false` to the `!something` style?

Comment (Soushi-Atsumi) [reply background/background.js]: Yes, I meant so. And as you said, I want you to change all cases to the `!something` style.

Comment (Joppe27) [reply background/background.js]: Applied in 19f4ad4.

**Pull request closed without merging**